File: app/src/main/java/com/mariotti/developer/futureclock/controllers/DatabaseAlarmController.java
package com.mariotti.developer.futureclock.controllers;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.mariotti.developer.futureclock.models.Alarm;
import com.mariotti.developer.futureclock.models.WeekDay;
import com.mariotti.developer.futureclock.models.database.AlarmBaseHelper;
import com.mariotti.developer.futureclock.models.database.AlarmCursorWrapper;
import com.mariotti.developer.futureclock.models.database.AlarmDbSchema.AlarmTable;
import com.mariotti.developer.futureclock.util.AlarmUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class DatabaseAlarmController {
    private static DatabaseAlarmController sDatabaseAlarmController;

    private SQLiteDatabase mDatabase;

    private DatabaseAlarmController(Context context) {
        mDatabase = new AlarmBaseHelper(context.getApplicationContext())
                .getWritableDatabase();
    }

    public static DatabaseAlarmController getDatabaseAlarmController(Context context) {
        if (sDatabaseAlarmController == null) {
            sDatabaseAlarmController = new DatabaseAlarmController(context);
        }
        return sDatabaseAlarmController;
    }

    public void addAlarm(Alarm alarm) throws Exception {
        if (alarm != null) {
            ContentValues values = getContentValues(alarm);

            mDatabase.beginTransaction();
            try {
                long rowID = mDatabase.insert(AlarmTable.NAME, null, values);
                if (rowID == -1) {
                    throw new Exception("Error inserting alarm");
                }
                mDatabase.setTransactionSuccessful();
            } finally {
                mDatabase.endTransaction();
            }
        }
    }

    private ContentValues getContentValues(Alarm alarm) {
        ContentValues values = new ContentValues();
        values.put(AlarmTable.Cols.UUID, alarm.getUuid().toString());
        String time = AlarmUtil.getHourAndMinuteAsString(alarm.getHour(), alarm.getMinute());
        values.put(AlarmTable.Cols.TIME, time);

        int days[] = alarm.getDays();
        values.put(AlarmTable.Cols.MONDAY, AlarmUtil.hasDay(days, WeekDay.MONDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.TUESDAY, AlarmUtil.hasDay(days, WeekDay.TUESDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.WEDNESDAY, AlarmUtil.hasDay(days, WeekDay.WEDNESDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.THURSDAY, AlarmUtil.hasDay(days, WeekDay.THURSDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.FRIDAY, AlarmUtil.hasDay(days, WeekDay.FRIDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.SATURDAY, AlarmUtil.hasDay(days, WeekDay.SATURDAY) ? 1 : 0);
        values.put(AlarmTable.Cols.SUNDAY, AlarmUtil.hasDay(days, WeekDay.SUNDAY) ? 1 : 0);

        values.put(AlarmTable.Cols.TIMEZONE, alarm.getTimezone());
        values.put(AlarmTable.Cols.ACTIVE, alarm.getActive() ? 1 : 0);
        return values;
    }

    public Alarm getAlarm(UUID id) {
        AlarmCursorWrapper cursorWrapper = queryAlarms(
                AlarmTable.Cols.UUID + " = ?",
                new String[]{id.toString()}
        );

        try {
            if (cursorWrapper.getCount() == 0) {
                return null;
            }

            cursorWrapper.moveToFirst();
            return cursorWrapper.getAlarm();
        } finally {
            cursorWrapper.close();
        }
    }

    private AlarmCursorWrapper queryAlarms(String whereClause, String[] whereArgs) {
        Cursor cursor = mDatabase.query(
                AlarmTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null
        );

        return new AlarmCursorWrapper(cursor);
    }

    public void updateAlarm(Alarm alarm) throws Exception {
        if (alarm != null) {
            String uuidString = alarm.getUuid().toString();
            ContentValues values = getContentValues(alarm);

            mDatabase.beginTransaction();
            try {
                int numRowsModified = mDatabase.update(
                        AlarmTable.NAME,
                        values,
                        AlarmTable.Cols.UUID + " = ?",
                        new String[]{uuidString}
                );
                if (numRowsModified != 1) {
                    throw new Exception("Error updating alarm " + uuidString);
                }
                mDatabase.setTransactionSuccessful();
            } finally {
                mDatabase.endTransaction();
            }
        }
    }

    public void deleteAlarm(UUID uuid) throws Exception {
        mDatabase.beginTransaction();
        try {
            int numRowsDeleted = mDatabase.delete(
                    AlarmTable.NAME,
                    AlarmTable.Cols.UUID + " = ?",
                    new String[]{uuid.toString()}
            );
            if (numRowsDeleted != 1) {
                throw new Exception("Error delete alarm " + uuid.toString());
            }
            mDatabase.setTransactionSuccessful();
        } finally {
            mDatabase.endTransaction();
        }
    }

    public List<Alarm> getAlarms() {
        AlarmCursorWrapper cursorWrapper = queryAlarms(null, null);
        List<Alarm> alarms = createListFromCursorWrapper(cursorWrapper);

        return alarms;
    }

    private List<Alarm> createListFromCursorWrapper(AlarmCursorWrapper cursorWrapper) {
        List<Alarm> alarms = new ArrayList<>();

        try {
            cursorWrapper.moveToFirst();
            while (!cursorWrapper.isAfterLast()) {
                alarms.add(cursorWrapper.getAlarm());
                cursorWrapper.moveToNext();
            }
        } finally {
            cursorWrapper.close();
        }

        return alarms;
    }

    public List<Alarm> getActiveAlarms() {
        AlarmCursorWrapper cursorWrapper = queryAlarms(
                AlarmTable.Cols.ACTIVE + " = 1",
                null
        );
        List<Alarm> alarms = createListFromCursorWrapper(cursorWrapper);

        return alarms;
    }
}

File: app/src/main/java/com/mariotti/developer/futureclock/controllers/DatabaseAlarmController.kt
package com.mariotti.developer.futureclock.controllers

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import com.mariotti.developer.futureclock.extensions.performTransaction
import com.mariotti.developer.futureclock.models.Alarm
import com.mariotti.developer.futureclock.models.WeekDay.FRIDAY
import com.mariotti.developer.futureclock.models.WeekDay.MONDAY
import com.mariotti.developer.futureclock.models.WeekDay.SATURDAY
import com.mariotti.developer.futureclock.models.WeekDay.SUNDAY
import com.mariotti.developer.futureclock.models.WeekDay.THURSDAY
import com.mariotti.developer.futureclock.models.WeekDay.TUESDAY
import com.mariotti.developer.futureclock.models.WeekDay.WEDNESDAY
import com.mariotti.developer.futureclock.models.database.AlarmBaseHelper
import com.mariotti.developer.futureclock.models.database.AlarmCursorWrapper
import com.mariotti.developer.futureclock.models.database.AlarmDbSchema.AlarmTable
import com.mariotti.developer.futureclock.util.getHourAndMinuteAsString
import com.mariotti.developer.futureclock.util.hasDay
import java.util.*

class DatabaseAlarmController private constructor(context: Context) {

    private val mDatabase: SQLiteDatabase

    init {
        mDatabase = AlarmBaseHelper(context.applicationContext).writableDatabase
    }

    companion object {
        private var instance: DatabaseAlarmController? = null

        fun getInstance(context: Context): DatabaseAlarmController {
            if (instance == null) {
                instance = DatabaseAlarmController(context)
            }

            return instance!!
        }
    }

    @Throws(Exception::class)
    fun addAlarm(alarm: Alarm): Unit {
        val values = getContentValues(alarm)

        mDatabase.performTransaction {
            val rowID = mDatabase.insert(AlarmTable.NAME, null, values)
            if (rowID == -1L) {
                throw Exception("Error inserting alarm")
            }
        }
    }

    private fun getContentValues(alarm: Alarm): ContentValues {
        val values = ContentValues()
        values.put(AlarmTable.Cols.UUID, alarm.uuid.toString())
        val time = getHourAndMinuteAsString(alarm.hour, alarm.minute)
        values.put(AlarmTable.Cols.TIME, time)

        val days = alarm.days
        values.put(AlarmTable.Cols.MONDAY, if (hasDay(days, MONDAY)) 1 else 0)
        values.put(AlarmTable.Cols.TUESDAY, if (hasDay(days, TUESDAY)) 1 else 0)
        values.put(AlarmTable.Cols.WEDNESDAY, if (hasDay(days, WEDNESDAY)) 1 else 0)
        values.put(AlarmTable.Cols.THURSDAY, if (hasDay(days, THURSDAY)) 1 else 0)
        values.put(AlarmTable.Cols.FRIDAY, if (hasDay(days, FRIDAY)) 1 else 0)
        values.put(AlarmTable.Cols.SATURDAY, if (hasDay(days, SATURDAY)) 1 else 0)
        values.put(AlarmTable.Cols.SUNDAY, if (hasDay(days, SUNDAY)) 1 else 0)

        values.put(AlarmTable.Cols.TIMEZONE, alarm.timezone)
        values.put(AlarmTable.Cols.ACTIVE, if (alarm.active) 1 else 0)
        return values
    }

    fun getAlarm(id: UUID): Alarm? {
        val cursorWrapper = queryAlarms(
                AlarmTable.Cols.UUID + " = ?",
                arrayOf(id.toString())
        )
        // use works like a try with resources
        cursorWrapper.use {
            if (it.count == 0) {
                return null
            }
            it.moveToFirst()
            return cursorWrapper.getAlarmFromDb()
        }
    }

    private fun queryAlarms(whereClause: String?, whereArgs: Array<String>?): AlarmCursorWrapper {
        val cursor = mDatabase.query(
                AlarmTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null)

        return AlarmCursorWrapper(cursor)
    }

    @Throws(Exception::class)
    fun updateAlarm(alarm: Alarm): Unit {
        val uuidString = alarm.uuid.toString()
        val values = getContentValues(alarm)

        mDatabase.performTransaction {
            val numRowsModified = mDatabase.update(
                    AlarmTable.NAME,
                    values,
                    AlarmTable.Cols.UUID + " = ?",
                    arrayOf(uuidString))
            if (numRowsModified != 1) {
                throw Exception("Error updating alarm " + uuidString)
            }
        }
    }

    @Throws(Exception::class)
    fun deleteAlarm(uuid: UUID): Unit {
        mDatabase.performTransaction {
            val numRowsDeleted = mDatabase.delete(
                    AlarmTable.NAME,
                    AlarmTable.Cols.UUID + " = ?",
                    arrayOf(uuid.toString()))
            if (numRowsDeleted != 1) {
                throw Exception("Error delete alarm " + uuid.toString())
            }
        }
    }

    fun getAlarms(): List<Alarm> {
        val cursorWrapper = queryAlarms(null, null)
        val alarms = createListFromCursorWrapper(cursorWrapper)
        return alarms
    }

    private fun createListFromCursorWrapper(cursorWrapper: AlarmCursorWrapper): List<Alarm> {
        val alarms = ArrayList<Alarm>()

        cursorWrapper.use {
            it.moveToFirst()
            while (!it.isAfterLast) {
                alarms.add(it.getAlarmFromDb())
                it.moveToNext()
            }
        }

        return alarms
    }

    fun getActiveAlarms(): List<Alarm> {
        val cursorWrapper = queryAlarms(
                AlarmTable.Cols.ACTIVE + " = 1",
                null)
        val alarms = createListFromCursorWrapper(cursorWrapper)
        return alarms
    }
}